apps/studio/components/interfaces/TableGridEditor/SidePanelEditor/SpreadsheetImport/SpreadsheetImport.utils.tsx

```diff
@@ -161,3 +170,57 @@ export const acceptedFileExtension = (file: any) => {
   const ext = file?.name.split('.').pop().toLowerCase()
   return UPLOAD_FILE_EXTENSIONS.includes(ext)
 }
+
+export function flagInvalidFileImport(file: File): boolean {
+  if (!file || !UPLOAD_FILE_TYPES.includes(file.type) || !acceptedFileExtension(file)) {
+    toast.error("Couldn't import file: only CSV files are accepted")
+    return true
+  } else if (file.size > MAX_TABLE_EDITOR_IMPORT_CSV_SIZE) {
+    toast.error(
+      <div className="space-y-1">
+        <p>The dashboard currently only supports importing of CSVs below 100MB.</p>
+        <p>For bulk data loading, we recommend doing so directly through the database.</p>
+        <Button asChild type="default" icon={<ExternalLink />} className="!mt-2">
+          <Link
+            href="https://supabase.com/docs/guides/database/tables#bulk-data-loading"
+            target="_blank"
+            rel="noreferrer"
+          >
+            Learn more
+          </Link>
+        </Button>
+      </div>,
+      { duration: Infinity }
+    )
+    return true
+  }
+
+  return false
+}
+
+// Custom event type for file drop processing
+const PROCESS_DROPPED_FILE_EVENT = 'supa::tableEditor::processDroppedFile' as const
+export interface ProcessDroppedFileEventDetail {
+  file: File
+}
+
+export interface ProcessDroppedFileEvent extends CustomEvent<ProcessDroppedFileEventDetail> {
+  type: typeof PROCESS_DROPPED_FILE_EVENT
+}
+
+// Helper function to dispatch the custom event
+export function dispatchProcessDroppedFileEvent(file: File): void {
```

Comment:
This is done with dispatched events rather than threading the file through the global state objects because passing the file through React state messes with later property accesses. I suspect that the Valtio proxy is messing with File prototype chain checks, but not sure.

Reply:
related discussion: https://github.com/pmndrs/valtio/discussions/739